Optimize GFPGAN and Improve User Experience

blessedcoolant opened the pull request:
This PR addresses the following issues.

1. GFPGAN is a post processing effect. Booting it at initialization meant that the model occupied VRAM that would otherwise be used for Stable Diffusion. This was resulting in systems having to use lower resolution on SD. As of this change, GFPGAN is no longer initialized at launch but rather just called in when a user prompts for it.
2. Loading GFPGAN at initialization meant booting torch modules in dream.py and then reinstance them when simplet2i.py is loaded again. As a side effect of this change, GFPGAN works off the torch modules loaded for SD and takes no performance hit in comparison to it being preloaded at initialization. Virtually no difference as far as I checked.
3. From a user experience perspective, the user now has absolute control on when to use GFPGAN without having to compromise VRAM by having it preloaded or having to restart the script to reinitialize if they didn't want to use GFPGAN or vice versa. Users can just pass **-G** along with their prompt to use GFPGAN.
4. Because this works on prompt basis now, you can have it on some, not have it on others. Whatever you like.

**NOTE**: I am not on my primary system. So I actually manually updated the code on Github. Someone double check that the formatting is correct before it is merged.

Thank you.

blessedcoolant commented:
Additionally, I'd also look into separating both GFPGAN and Real-ESRGAN. GFPGAN is for blind face restoration and Real-ESRGAN is for upscaling. I think it'd be better for users to have the option to access these individually so they can use GFPGAN with Real-ESRGAN on images with faces and use just Real-ESRGAN for images that just need upscaling.

blessedcoolant commented:
Don't merge this yet please. I'm making some important changes.

blessedcoolant commented:
I've made some further improvements.

Instead of running GFPGAN while still processing SD and leading to a ton of memory issues, the app now runs GFPGAN after SD's images have been generated.

While SD is running, the app generates images and previews for the user to see their progress. When the processing ends, the app rescales all of the generated images to an upscaled version using GFPGAN and Real-ESRGAN.

This way no matter how big the batch size gets or samples get, there will likely not be any memory or CUDA crashes.

I've tested it out quite a bit. Seems to be working great. Let me know if you want any changes. @lstein 

Should be good to pull now if u wish.

lstein commented:
Hi Folks, will inspect and merge this shortly. This project has grown quite a bit so I'm going to start to do formal feature branches, bugfix branches, etc. I don't want users to get frustrated.

blessedcoolant commented:
> Hi Folks, will inspect and merge this shortly. This project has grown quite a bit so I'm going to start to do formal feature branches, bugfix branches, etc. I don't want users to get frustrated.

Thank you. Will keep an eye out for it. No issues. It's the right thing to do if this repo is going to grow larger.

Oceanswave commented:
Unable to get this to run - each iteration says GFPGAN not initialized, it must be loaded via the --gfpgan argument  when specifying the -G argument

Oceanswave commented:
Also, why iterate through the images twice when it's already happening after each image is generated on line 321?

Oceanswave commented on ldm/simplet2i.py:
```diff
@@ -323,6 +336,21 @@ def process_image(image,seed):
                             image_callback(image, seed)
                     seed = self._new_seed()
 
+                if gfpgan_strength > 0:
```
This never gets run

blessedcoolant replied:
Are you passing GFPGAN as a prompt argument?

`<prompt> -G 0.5`

Oceanswave commented:
It doesn't report any error, just the message that it can't be initialized and it fails, I get the non gfpgan'd image in my output

blessedcoolant commented:
> It doesn't report any error, just the message that it can't be initialized and it fails, I get the non gfpgan'd image in my output

Unsure. It's hard to tell without more info. I just made a clean pull from github and merged with mine. Seems to be working fine.

Oceanswave commented:
ok, I created a new conda environment and activated that, probably a me issue from testing the installation yesterday

My time for
```
a man wearing a pineapple hat -n6 -S1 -G1
```

is now twice over as long and it outputs that it processed 12 files instead of 6:

```
D:\stable-diffusion\000014.1902099158.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1902099158
D:\stable-diffusion\000014.3697446297.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3697446297
D:\stable-diffusion\000014.1077098826.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1077098826
D:\stable-diffusion\000014.1212597026.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1212597026
D:\stable-diffusion\000014.3743693958.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3743693958
D:\stable-diffusion\000014.2939900369.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2939900369
D:\stable-diffusion\000014.1902099158.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1902099158
D:\stable-diffusion\000014.3697446297.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3697446297
D:\stable-diffusion\000014.1077098826.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1077098826
D:\stable-diffusion\000014.1212597026.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1212597026
D:\stable-diffusion\000014.3743693958.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3743693958
D:\stable-diffusion\000014.2939900369.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2939900369
```

So that seems to be bad, also my VRAM usage is now worse after the first run because it seems to keep it gpfgan around until the GC gets to it, so I have to wait around for it to clear.

lstein commented:
I haven't even gotten that far. The problem I'm having with this PR is that the --gfpgan command line switch is missing from the argv parsing list. Was it intended that gfpgan be activated by specifying --gfpgan_strength when starting the script?

lstein commented:
Ok, I think this is still a work in progress and will convert into a draft. I'm going to branch the current version (with Oceanswave second PR) into a feature branch so that people can continue to work with it.

blessedcoolant commented:
> ok, I created a new conda environment and activated that, probably a me issue from testing the installation yesterday
> 
> My time for
> 
> ```
> a man wearing a pineapple hat -n6 -S1 -G1
> ```
> 
> is now twice over as long and it outputs that it processed 12 files instead of 6:
> 
> ```
> D:\stable-diffusion\000014.1902099158.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1902099158
> D:\stable-diffusion\000014.3697446297.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3697446297
> D:\stable-diffusion\000014.1077098826.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1077098826
> D:\stable-diffusion\000014.1212597026.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1212597026
> D:\stable-diffusion\000014.3743693958.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3743693958
> D:\stable-diffusion\000014.2939900369.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2939900369
> D:\stable-diffusion\000014.1902099158.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1902099158
> D:\stable-diffusion\000014.3697446297.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3697446297
> D:\stable-diffusion\000014.1077098826.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1077098826
> D:\stable-diffusion\000014.1212597026.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1212597026
> D:\stable-diffusion\000014.3743693958.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3743693958
> D:\stable-diffusion\000014.2939900369.png: "a man wearing a pineapple hat" -s1 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2939900369
> ```
> 
> So that seems to be bad, also my VRAM usage is now worse after the first run because it seems to keep it gfpgan around until the GC gets to it, so I have to wait around for it to clear.
> 
> This seems to be an extremely bad user experience. It gives me a _TON_ of errors if I didn't set my gfpgan up correctly the first time (which I didn't since I switched conda environments).

The GFPGAN implementation is yours. I did nothing there. If anything, I got rid of extra user warnings that are from the obsolete GFPGAN code.

Even in your code, you were generating 12 images -- 6 from SD and 6 Upscaled. The only difference is that I am materializing them to the disk so the user has a preview.

I am not sure what issues you are facing with VRAM usage. For me the load has been far better because you are no longer running GFPGAN and SD at the same time. In your implementation,. 512x768 which easily renders on my system kept crashing because GFPGAN started to hog the memory while the SD tensors were still in loop.

The GC collection remains the same.

To be very honest, I think this is a more ideal and user friendly way to do it. Because instead of going through the hoops of setting up a bunch of starter commands, a user can just trigger and launch GFPGAN with the help of a prompt which is the most natural way to do it.

If you hate seeing the log of images being saved, it's a one line fix to do that.

Either way, I'll let @lstein decide which one he prefers. Personally I think I am staying with mine because I truly believe it consumes far less memory overload and has a much nicer way of doing upscaling through prompts.

blessedcoolant commented:
> I haven't even gotten that far. The problem I'm having with this PR is that the --gfpgan command line switch is missing from the argv parsing list. Was it intended that gfpgan be activated by specifying --gfpga

You no longer need to use --gfpgan with this method. You can use the other commands to change the locations of the GFPGAN folders but otherwise, the entire process not works directly from the prompt that depends on the value of the gfpgan_strength. It's intended.

Why have two different start up modes for the script when the base script can handle the same thing through prompts right?

lstein commented:
Everyone, thank you all for your efforts on this. Further work on gfpgan should be done on a new branch named **gfpgan-development**. From the maintenance point of view, I'd like to see the majority of the gfpgan code moved into its own module located under ldm/dream/ so that it isn't cluttering the already busy dream.py and simplet2i.py modules.

Oceanswave commented:
> > So that seems to be bad, also my VRAM usage is now worse after the first run because it seems to keep it gfpgan around until the GC gets to it, so I have to wait around for it to clear.
> > This seems to be an extremely bad user experience. It gives me a _TON_ of errors if I didn't set my gfpgan up correctly the first time (which I didn't since I switched conda environments).
> 
> The GFPGAN implementation is yours. I did nothing there. If anything, I got rid of extra user warnings that are from the obsolete GFPGAN code.

But we talked about this, another loop has been introduced and it's running _run_gfpgan for each image twice.

> 
> Even in your code, you were generating 12 images -- 6 from SD and 6 Upscaled. The only difference is that I am materializing them to the disk so the user has a preview.
> 
> I am not sure what issues you are facing with VRAM usage. For me the load has been far better because you are no longer running GFPGAN and SD at the same time. In your implementation,. 512x768 which easily renders on my system kept crashing because GFPGAN started to hog the memory while the SD tensors were still in loop.
> 
> The GC collection remains the same.

But it's not, since every image it loads and unloads a new instance of GFPGAN


> To be very honest, I think this is a more ideal and user friendly way to do it. Because instead of going through the hoops of setting up a bunch of starter commands, a user can just trigger and launch GFPGAN with the help of a prompt which is the most natural way to do it.

I'm a big fan of failing fast - if a user started with --gfpgan, we know apriori that we should test for gfpgan dependencies before failing at runtime

lstein commented:
> You no longer need to use --gfpgan with this method. You can use the other commands to change the locations of the GFPGAN folders but otherwise, the entire process now works directly from the prompt that depends on the value of the gfpgan_strength. It's intended.

Ah, OK. The README needs to be updated because it still refers to --gfpgan on the command line. In any case, let's continue this on the new feature branch. I'm really impressed with what GFPGAN can do!

blessedcoolant commented:
> But we talked about this, another loop has been introduced and it's running _run_gfpgan for each image twice.

No it is not. It is running GFPGAN only once per image. And only when the user has particularly requested for it. You got this idea of double generation because I was saving both the SD and GFP images. They were already being generated. The only difference between your patch and mine is that they're no longer happening simultaneously coz that was leading to a multitude of memory issues.

blessedcoolant commented:
> > But we talked about this, another loop has been introduced and it's running _run_gfpgan for each image twice.
> 
> No it is not. It is running GFPGAN only once per image. And only when the user has particularly requested for it. You got this idea of double generation because I was saving both the SD and GFP images. They were already being generated. The only difference between your patch and mine is that they're no longer happening simultaneously coz that was leading to a multitude of memory issues.
> 
> Particularly this issue: #112 -- I tackled this making GFPGAN perform its upscaling after SD has done its work. If you do not do this, the second iteration will crash because there is no longer enough memory for both GFPGAN and SD to run at the same time. You can try this yourself.

blessedcoolant commented:
> Ah, OK. The README needs to be updated because it still refers to --gfpgan on the command line. In any case, let's continue this on the new feature branch. I'm really impressed with what GFPGAN can do!

My PR has an updated Readme on that. Let me know if I need to expand on it.

Oceanswave commented:
> No it is not. It is running GFPGAN only once per image. And only when the user has particularly requested for it. You got this idea of double generation because I was saving both the SD and GFP images. They were already being generated. The only difference between your patch and mine is that they're no longer happening simultaneously coz that was leading to a multitude of memory issues.
> 
> Particularly this issue: #112 -- I tackled this making GFPGAN perform its upscaling after SD has done its work. If you do not do this, the second iteration will crash because there is no longer enough memory for both GFPGAN and SD to run at the same time.

call #1: https://github.com/blessedcoolant/stable-diffusion/blob/9ff140058b0b02be47495389076de6097caeae97/ldm/simplet2i.py#L326
call #2: https://github.com/blessedcoolant/stable-diffusion/blob/9ff140058b0b02be47495389076de6097caeae97/ldm/simplet2i.py#L344

blessedcoolant commented:
> > No it is not. It is running GFPGAN only once per image. And only when the user has particularly requested for it. You got this idea of double generation because I was saving both the SD and GFP images. They were already being generated. The only difference between your patch and mine is that they're no longer happening simultaneously coz that was leading to a multitude of memory issues.
> > Particularly this issue: #112 -- I tackled this making GFPGAN perform its upscaling after SD has done its work. If you do not do this, the second iteration will crash because there is no longer enough memory for both GFPGAN and SD to run at the same time.
> 
> call #1: https://github.com/blessedcoolant/stable-diffusion/blob/9ff140058b0b02be47495389076de6097caeae97/ldm/simplet2i.py#L326 call #2: https://github.com/blessedcoolant/stable-diffusion/blob/9ff140058b0b02be47495389076de6097caeae97/ldm/simplet2i.py#L344
> 
> hack to get the count to be correct, but the callback is still called for each image 2x
> 
> https://github.com/blessedcoolant/stable-diffusion/blob/9ff140058b0b02be47495389076de6097caeae97/ldm/simplet2i.py#L367

Fuck. Seems like I updated the wrong code. Here's how it is on my end. Let me fix it up.

![Code_oEcqcXdaii](https://user-images.githubusercontent.com/54517381/186951301-2d91b4a1-edcf-46be-b8cc-e01f202d6a99.png)

blessedcoolant commented:
@lstein @Oceanswave I made the right pull now. Hopefully that's better. Sorry for the confusion lol. Now I hope it makes more sense.

Oceanswave commented:
> Particularly this issue: #112 -- I tackled this making GFPGAN perform its upscaling after SD has done its work. If you do not do this, the second iteration will crash because there is no longer enough memory for both GFPGAN and SD to run at the same time.

Yeah, but now I run into the OOM issue with 24GB ram after an iteration of 10, so we probably need a gc.collect() in there or a torch.cuda().empty_cache() since it's doing a bunch more memory allocations.

Oceanswave commented:
> @lstein @Oceanswave I made the right pull now. Hopefully that's better. Sorry for the confusion lol. Now I hope it makes more sense.

cool, lemmie pull

Oceanswave commented:
> @lstein @Oceanswave I made the right pull now. Hopefully that's better. Sorry for the confusion lol. Now I hope it makes more sense.

ah, that generated_iterations var is still messing it up

blessedcoolant commented:
> > @lstein @Oceanswave I made the right pull now. Hopefully that's better. Sorry for the confusion lol. Now I hope it makes more sense.
> 
> ah, that generated_iterations var is still messing it up
> 
> oh, and it's still calling the callback twice, so the file gets bounced multiple times, this is reflected in the logs too

Messing it up in what sense?

I am calling the callback twice intentionally. Because if I don't do that, then the user has to wait until after SD to actually get an image preview. If they want to cancel it, they have to wait for the entire SD iterations to end before they get any kind of feedback on what kind if images are being generated.

The first callback outputs the raw SD images. So this will act as user feedback while SD finishes the job. And if the user is okay with it, they can just let it be and once SD has completed, GFPGAN will upscale and replace the older images with newer ones.

The logs only show the resulting outputs. Previously we only saved GFPGAN to disk so it only showed those. But it's an easy fix. Can just get them to stop counting the base image.

But overall the number of images actually generated (not saved to disk) but rather created by GFPGAN and SD remain the same. If you do 6 images, there will be 12 outputs because GFFGAN generates an entirely new image too.

Oceanswave commented:
> Messing it up in what sense?

Traceback (most recent call last):
  File "./scripts/dream.py", line 434, in <module>
    main()
  File "./scripts/dream.py", line 36, in main
    from ldm.simplet2i import T2I
  File "c:\src\stable-diffusion-updream\ldm\simplet2i.py", line 313
    generated_iterations = []
    
> I am calling the callback twice intentionally. Because if I don't do that, then the user has to wait until after SD to actually get an image preview. If they want to cancel it, they have to wait for the entire SD iterations to end before they get any kind of feedback on what kind if images are being generated.

Yeah, but one of the sort of implicit things is to have that callback be executed a single time per requested image - in my case I have a callback that saves images into a redis cache, I don't want to invoke that network traffic twice for every image, when I know I wanted the fixed/upsampled version since I specified -G -- if I wanted to do a comparison I could just resubmit without the flag, or do the new -S-1 feature that @lstein added

plus since gfpgan takes less than a second to run without it being reloaded I mean what is really being saved here. I can't click the file and click it again in that amount of time. or stop the process

blessedcoolant commented:
> > Messing it up in what sense?
> 
> Traceback (most recent call last): File "./scripts/dream.py", line 434, in main() File "./scripts/dream.py", line 36, in main from ldm.simplet2i import T2I File "c:\src\stable-diffusion-updream\ldm\simplet2i.py", line 313 generated_iterations = []
> 
> > I am calling the callback twice intentionally. Because if I don't do that, then the user has to wait until after SD to actually get an image preview. If they want to cancel it, they have to wait for the entire SD iterations to end before they get any kind of feedback on what kind if images are being generated.
> 
> Yeah, but one of the sort of implicit things is to have that callback be executed a single time per requested image - in my case I have a callback that saves images into a redis cache, I don't want to invoke that network traffic twice for every image, when I know I wanted the fixed/upsampled version since I specified -G -- if I wanted to do a comparison I could just resubmit without the flag, or do the new -S-1 feature that @lstein added
> 
> plus since gfpgan takes less than a second to run without it being reloaded I mean what is really being saved here. I can't click the file and click it again in that amount of time. or stop the process

So do you reckon we don't have the original images and just do the callback once?

Oceanswave commented:
> So do you reckon we don't have the original images and just do the callback once?

that's my thinking, so that commands are idempotent - following this pattern, I'd expect a gobig implementation to not bounce the resulting image n times, but rather just output the final scaled image. it sort of breaks a level of idempotency if a callback needs to be created in a way that handles the side-effects of the command

blessedcoolant commented:
> > So do you reckon we don't have the original images and just do the callback once?
> 
> that's my thinking, so that commands are idempotent - following this pattern, I'd expect a gobig implementation to not bounce the resulting image n times, but rather just output the final scaled image. it sort of breaks a level of idempotency if a callback needs to be created in a way that handles the side-effects of the command

One of the reasons I went for this was because in large operations such as many iterations, there won't be any image feedback for a while. It's okay for smaller generations though coz like you said the GFPGAN thing happens almost right away. But it's only an issue when the processing takes a while in large batches.

Oceanswave commented:
So with it working more intended, and comparing this side-by-side, I get a ~30% reduction in speed with this:

original:
```
5 images generated in 33.04s
Outputs:
D:\stable-diffusion\000018.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
D:\stable-diffusion\000018.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
D:\stable-diffusion\000018.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
D:\stable-diffusion\000018.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
D:\stable-diffusion\000018.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
```

new:
```
5.0 images generated in 43.43s
Outputs:
D:\stable-diffusion\000017.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
D:\stable-diffusion\000017.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
D:\stable-diffusion\000017.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
D:\stable-diffusion\000017.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
D:\stable-diffusion\000017.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
D:\stable-diffusion\000017.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
D:\stable-diffusion\000017.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
D:\stable-diffusion\000017.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
D:\stable-diffusion\000017.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
D:\stable-diffusion\000017.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
```

That's not great - I also sill get the oom on larger images/batches

Oceanswave commented:
> One of the reasons I went for this was because in large operations such as many iterations, there won't be any image feedback for a while. It's okay for smaller generations though coz like you said the GFPGAN thing happens almost right away. But it's only an issue when the processing takes a while in large batches.

well, with the orig. implementation with it happening on the first pass of the iterator, it actually happens immediately after the image is generated, thus one doesn't need to wait

lstein commented:
I hate waiting for initialization, so I'd rather get it all over at the
beginning. It's particularly a problem on my system because its an HPC
using a shared NFS mount. So loading big models and the modules themselves
really makes me wait. In any case, I'm going to be running errands this
afternoon, but after that I'll put the new commits through their paces and
report my impressions back.

Lincoln

On Fri, Aug 26, 2022 at 12:17 PM blessedcoolant ***@***.***>
wrote:

> I haven't even gotten that far. The problem I'm having with this PR is
> that the --gfpgan command line switch is missing from the argv parsing
> list. Was it intended that gfpgan be activated by specifying --gfpga
>
> You no longer need to use --gfpgan with this method. You can use the other
> commands to change the locations of the GFPGAN folders but otherwise, the
> entire process not works directly from the prompt that depends on the value
> of the gfpgan_strength. It's intended.
>
> Why have two different start up modes for the script when the base script
> can handle the same thing through prompts right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lstein/stable-diffusion/pull/102#issuecomment-1228686622>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3EVKEKKOFXEIIKG5GDHLV3DUYDANCNFSM57WGQHAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

*Lincoln Stein*

Head, Adaptive Oncology, OICR

Senior Principal Investigator, OICR

Professor, Department of Molecular Genetics, University of Toronto



Tel: 416-673-8514

Cell: 416-817-8240

***@***.***



*E**xecutive Assistant*

Michelle Xin

Tel: 647-260-7927

***@***.*** ***@***.***>*



*Ontario Institute for Cancer Research*

MaRS Centre, 661 University Avenue, Suite 510, Toronto, Ontario, Canada M5G
0A3

@OICR_news
<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftwitter.com%2Foicr_news&data=04%7C01%7CMichelle.Xin%40oicr.on.ca%7C9fa8636ff38b4a60ff5a08d926dd2113%7C9df949f8a6eb419d9caa1f8c83db674f%7C0%7C0%7C637583553462287559%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PS9KzggzFoecbbt%2BZQyhkWkQo9D0hHiiujsbP7Idv4s%3D&reserved=0>
 | www.oicr.on.ca



*Collaborate. Translate. Change lives.*



This message and any attachments may contain confidential and/or privileged
information for the sole use of the intended recipient. Any review or
distribution by anyone other than the person for whom it was originally
intended is strictly prohibited. If you have received this message in
error, please contact the sender and delete all copies. Opinions,
conclusions or other information contained in this message may not be that
of the organization.

lstein commented:
Another thing to keep in mind is that the user may hit ^C during a
long-running set of iterations. Up until the interrupt hits, you want all
the previously-generated images to be complete, processed, and logged.

Lincoln

On Fri, Aug 26, 2022 at 1:05 PM blessedcoolant ***@***.***>
wrote:

> So do you reckon we don't have the original images and just do the
> callback once?
>
> that's my thinking, so that commands are idempotent - following this
> pattern, I'd expect a gobig implementation to not bounce the resulting
> image n times, but rather just output the final scaled image. it sort of
> breaks a level of idempotency if a callback needs to be created in a way
> that handles the side-effects of the command
>
> One of the reasons I went for this was because in large operations such as
> many iterations, there won't be any image feedback for a while. It's okay
> for smaller generations though coz like you said the GFPGAN thing happens
> almost right away. But it's only an issue when the processing takes a while
> in large batches.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lstein/stable-diffusion/pull/102#issuecomment-1228728188>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3EVN22F5WVCOLO54MI7DV3D2N7ANCNFSM57WGQHAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

*Lincoln Stein*

Head, Adaptive Oncology, OICR

Senior Principal Investigator, OICR

Professor, Department of Molecular Genetics, University of Toronto



Tel: 416-673-8514

Cell: 416-817-8240

***@***.***



*E**xecutive Assistant*

Michelle Xin

Tel: 647-260-7927

***@***.*** ***@***.***>*



*Ontario Institute for Cancer Research*

MaRS Centre, 661 University Avenue, Suite 510, Toronto, Ontario, Canada M5G
0A3

@OICR_news
<https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftwitter.com%2Foicr_news&data=04%7C01%7CMichelle.Xin%40oicr.on.ca%7C9fa8636ff38b4a60ff5a08d926dd2113%7C9df949f8a6eb419d9caa1f8c83db674f%7C0%7C0%7C637583553462287559%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PS9KzggzFoecbbt%2BZQyhkWkQo9D0hHiiujsbP7Idv4s%3D&reserved=0>
 | www.oicr.on.ca



*Collaborate. Translate. Change lives.*



This message and any attachments may contain confidential and/or privileged
information for the sole use of the intended recipient. Any review or
distribution by anyone other than the person for whom it was originally
intended is strictly prohibited. If you have received this message in
error, please contact the sender and delete all copies. Opinions,
conclusions or other information contained in this message may not be that
of the organization.

blessedcoolant commented:
> So with it working more intended, and comparing this side-by-side, I get a ~30% reduction in speed with this:
> 
> original:
> 
> ```
> 5 images generated in 33.04s
> Outputs:
> D:\stable-diffusion\000018.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
> D:\stable-diffusion\000018.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
> D:\stable-diffusion\000018.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
> D:\stable-diffusion\000018.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
> D:\stable-diffusion\000018.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
> ```
> 
> new:
> 
> ```
> 5.0 images generated in 43.43s
> Outputs:
> D:\stable-diffusion\000017.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
> D:\stable-diffusion\000017.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
> D:\stable-diffusion\000017.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
> D:\stable-diffusion\000017.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
> D:\stable-diffusion\000017.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
> D:\stable-diffusion\000017.1.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S1
> D:\stable-diffusion\000017.577090037.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S577090037
> D:\stable-diffusion\000017.896706486.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S896706486
> D:\stable-diffusion\000017.3829394925.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S3829394925
> D:\stable-diffusion\000017.2346590147.png: "a man wearing a pineapple hat" -s50 -b1 -W512 -H512 -C7.5 -mk_euler -G1.0 -F -S2346590147
> ```
> 
> That's not great - I also sill get the oom on larger images/batches

The inference time is larger because its processing GFPGAN sequentially but doing this lets you do larger resolutions and more batches than you do otherwise. There's a limit to that but overall it might be worth the trade off imo.

blessedcoolant commented:
> Another thing to keep in mind is that the user may hit ^C during a long-running set of iterations. Up until the interrupt hits, you want all the previously-generated images to be complete, processed, and logged. Lincoln
> […](#)
> On Fri, Aug 26, 2022 at 1:05 PM blessedcoolant ***@***.***> wrote: So do you reckon we don't have the original images and just do the callback once? that's my thinking, so that commands are idempotent - following this pattern, I'd expect a gobig implementation to not bounce the resulting image n times, but rather just output the final scaled image. it sort of breaks a level of idempotency if a callback needs to be created in a way that handles the side-effects of the command One of the reasons I went for this was because in large operations such as many iterations, there won't be any image feedback for a while. It's okay for smaller generations though coz like you said the GFPGAN thing happens almost right away. But it's only an issue when the processing takes a while in large batches. — Reply to this email directly, view it on GitHub <[#102 (comment)](https://github.com/lstein/stable-diffusion/pull/102#issuecomment-1228728188)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAA3EVN22F5WVCOLO54MI7DV3D2N7ANCNFSM57WGQHAQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>
> -- *Lincoln Stein* Head, Adaptive Oncology, OICR Senior Principal Investigator, OICR Professor, Department of Molecular Genetics, University of Toronto Tel: 416-673-8514 Cell: 416-817-8240 ***@***.*** *E**xecutive Assistant* Michelle Xin Tel: 647-260-7927 ***@***.*** ***@***.***>* *Ontario Institute for Cancer Research* MaRS Centre, 661 University Avenue, Suite 510, Toronto, Ontario, Canada M5G 0A3 @OICR_news <https://can01.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftwitter.com%2Foicr_news&data=04%7C01%7CMichelle.Xin%40oicr.on.ca%7C9fa8636ff38b4a60ff5a08d926dd2113%7C9df949f8a6eb419d9caa1f8c83db674f%7C0%7C0%7C637583553462287559%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=PS9KzggzFoecbbt%2BZQyhkWkQo9D0hHiiujsbP7Idv4s%3D&reserved=0> | [www.oicr.on.ca](http://www.oicr.on.ca) *Collaborate. Translate. Change lives.* This message and any attachments may contain confidential and/or privileged information for the sole use of the intended recipient. Any review or distribution by anyone other than the person for whom it was originally intended is strictly prohibited. If you have received this message in error, please contact the sender and delete all copies. Opinions, conclusions or other information contained in this message may not be that of the organization.

Agreed. The current implementation has all the images logged at break time. Either if its the regular SD version or the upscaled.. depending on what part of the process we're at.

>I hate waiting for initialization, so I'd rather get it all over at the
beginning. It's particularly a problem on my system because its an HPC
using a shared NFS mount. So loading big models and the modules themselves
really makes me wait. In any case, I'm going to be running errands this
afternoon, but after that I'll put the new commits through their paces and
report my impressions back.

There's virtually no different for inference of GFPGAN whether we load it initially or load it during. The only difference is the memory. Having GFPGAN in memory when not being used is not worth it imo.

blessedcoolant commented:
> > One of the reasons I went for this was because in large operations such as many iterations, there won't be any image feedback for a while. It's okay for smaller generations though coz like you said the GFPGAN thing happens almost right away. But it's only an issue when the processing takes a while in large batches.
> 
> well, with the orig. implementation with it happening on the first pass of the iterator, it actually happens immediately after the image is generated, thus one doesn't need to wait

If you do try to do multiple iterations on that, it'll fail on the second as others have reported too.

I think at this stage, I've implemented the best version I think in regards to this. I'm open to ideas but of the two versions we have now, I prefer mine personally.

Not sure if there's anything else I can add to this PR for now.

Oceanswave commented:
> Agreed. The current implementation has all the images logged at break time. Either if its the regular SD version or the upscaled.. depending on what part of the process we're at.

But if I say -n5 -G1, and I ^C after the 3rd, because I saw the first and I got want I wanted in terms of content, I don't have the upscaled version at that point, so the operation isn't really 'complete' at that point. on the main branch, if run and see an image on the file system it's done done. no takebacksies.

> There's virtually no difference for inference of GFPGAN whether we load it initially or load it during. The only difference is the memory. Having GFPGAN in memory when not being used is not worth it imo.

Say I have a slow network share that my code is on, that I'm mounting. for every image it's making a call to see if gfpgan exists, loading the model (it could have changed) across the network, then getting it into VRAM. processing the image, setting gfpgan to None, which lets the GC free it on the next cycle.

Probably what I'm hitting on this PR is things are getting loaded faster than the GC is able to get to them and I eventually get an OOM

Oceanswave commented:
> I think at this stage, I've implemented the best version I think in regards to this. I'm open to ideas but of the two versions we have now, I prefer mine personally.
> 
> Not sure if there's anything else I can add to this PR for now.

Yeah, I understand the memory allocation issue - I think there's a middle ground here, and some data needs to be gathered to really determine an approach.

We might just need to sprinkle more gc.collect() around, or, the default mode as you indicate is to unload it, but also have a mode where it keeps it - solving the perf vs mem problem depending on the scenario.

blessedcoolant commented:
>  I don't have the upscaled version at that point, so the operation isn't really 'complete' at that point. on the main branch, if run and see an image on the file system it's done done. no takebacksies.

For stand out cases like that what if we provide an option to just `--upscale` any image they like?

> Probably what I'm hitting on this PR is things are getting loaded faster than the GC is able to get to them and I eventually get an OOM

Using my method I haven't hit an OOM yet. Can you tell me what your testing parameters are for that? I'll try to replicate that and see if it gets me an OOM too. As long as I stick to this process, I'm able to generate seemingly infinite images. I just did 350 and it worked fine. Could probably leave it on for a few hours and see if it crashes at some point.

> We might just need to sprinkle more gc.collect() around, or, the default mode as you indicate is to unload it, but also have a mode where it keeps it - solving the perf vs mem problem depending on the scenario.

Wish we had some easier way to get some user feedback on what people would prefer seeing.

I'm exploring some more ideas to see if they're any better than what we have at the moment. I'll hit back here if I can work something out.

lstein commented:
Slightly off topic, but is there a way to get both GPUs going during an
iteration run on a multi-GPU system? That would be a major benefit for a
few lucky people.

I’m taking a break from code review and testing tonight, and will weigh in
on the GFPGAN controversy tomorrow.

Lincoln

On Fri, Aug 26, 2022 at 1:37 PM Sean McLellan ***@***.***>
wrote:

> Agreed. The current implementation has all the images logged at break
> time. Either if its the regular SD version or the upscaled.. depending on
> what part of the process we're at.
>
> But if I say -n5 -G1, and I ^C after the 3rd, because I saw the first and
> I got want I wanted in terms of content, I don't have the upscaled version
> at that point, so the operation isn't really 'complete' at that point. on
> the main branch, if run and see an image on the file system it's done done.
> no takebacksies.
>
> There's virtually no difference for inference of GFPGAN whether we load it
> initially or load it during. The only difference is the memory. Having
> GFPGAN in memory when not being used is not worth it imo.
>
> Say I have a slow network share that my code is on, that I'm mounting. for
> every image it's making a call to see if gfpgan exists, loading the model
> (it could have changed) across the network, then getting it into VRAM.
> processing the image, setting gfpgan to None, which lets the GC free it on
> the next cycle.
>
> Probably what I'm hitting on this PR is things are getting loaded faster
> than the GC is able to get to them and I eventually get an OOM
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lstein/stable-diffusion/pull/102#issuecomment-1228753776>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA3EVPIU7XAQ3VKTFZ2Q23V3D6FZANCNFSM57WGQHAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Written on my cell phone. Anything that seems odd is the fault of
auto-correct.

blessedcoolant commented:
So I made some changes on my end. Here's what I settled with.

## Patch 1

Both GFPGAN and Real-ESRGAN are now independent of each other. Rather than triggering both each time, users can now control them independently depending on the kind of their prompt.

To support this, I added two new prompt arguments.

- `-U --upscale`

Supplying the -U prompt argument will trigger Real-ESRGAN Image Scaling (that was running currently with GFPGAN). This will NOT perform Face Restoration. That is the job of GFPGAN so we should let it do that.

`--upscale` takes two values. The first value controls the scale factor for Real-ESRGAN. **I implemented the 4x version of Real-ESRGAN** too. You can pass either `2` or `4` as the first value. The second value controls the intensity of the scaling. Generally these AI's really smooth out the image removing a lot of texture. With this value, you can get considerably better looking results. Ideal values of around `0.5 to 0.8`.

- `-G --gfpgan_strength`

Supplying the -G prompt argument will trigger GFPGAN. This exclusively performs face restoration. It no longer performs scaling.

`--gfpgan_strength` takes one value which controls the intensity of the face restoration that is being applied. Ideal values are around `0.5 to 0.8`.

- Users can independently call either -G or -U to either upscale or do face restoration.
- Users can call both -U and -G to perform upscaling and also face restoration. When you call both, the upscaling is done first and then the face restoration. This ensures that the face restoration module has the most data to work with leading to the best results for the face quality.
- `-save_orig` : A new prompt argument that allows you to save original images when performing upscaling or face restoration. Use this in tandem with -U or -G or both. This way you can save and compare the original SD generation with the artificially enhanced versions if need be.
- Moved GFPGAN and Real_ESRGAN from the simple2ti.py to another folder - **ldm/gfpgan** -- Cleaned up simple2ti.py accordingly.
- GFPGAN and Real-ESRGAN now log which file is being upscaled instead of the generic message so users can keep track of the progress.

---

# Memory Issues

After trying out a ton of options, I have to once again get back to the point that my implementation of the pipeline is the most ideal one in regards to GFPGAN and Real-ESRGAN at this current point of time.

The main goal is to create beautiful images and hopefully cater to systems even with lower VRAM so the maximum number of people can use it.

With that in mind, I have to reaffirm on my feeling that to avoid any memory issues, GFPGAN and Real ESRGAN will be performed after SD has completed image generation. Performing them while SD is still in memory will inevitably crash the generation sooner than later.

I have generated over 1500 images since we last spoke with the new system while scaling all of them by 2X and performing face restoration. I have not had a single crash by OOM so far.

The big issue of a user canceling a generation to not have an upscaled result is no longer a problem. If a user cancels a generation, they will have all images generated until that point. If they like any of those images, they can just get the seed from it and repass it to the generator along with -U and -G to perform scaling and face restoration if they want to. Problem solved.

---

This is what I have so far. I am trying to make the whole process even better. If I come up with any more ideas, I'll post here. But if you guys agree with this, I think I might make a new PR for that in a day or two with all the changes I mentioned above and any others that I might add.

---

@lstein I will have a look into it. Maybe there are ways to defer them to a secondary GPU. But like you said, it'll be handy only for a lucky few .. but no reason not to implement it if we can.

blessedcoolant commented:
@lstein @Oceanswave I've pulled the Patch 1 detailed above and all the changes that I made with it. Completed the documentation too.

lstein commented:
Thanks for all the work on this! I am testing now after having taken a break for a family event.

blessedcoolant commented:
> Thanks for all the work on this! I am testing now after having taken a break for a family event.

No rush mate. :) Take your time. Thank you for your time and effort too.

lstein commented:
@blessedcoolant Is there any particular reason you changed the default image height to 768?

blessedcoolant commented:
> @blessedcoolant Is there any particular reason you changed the default image height to 768?

Oh crap. That's just my defaults for my own system so I don't have to keep passing it. I forgot to reset it to the before I made the push. You can just change it back.

lstein commented:
@blessedcoolant I've run into another problem. For benchmarking I'm running 5 iterations with the same starting seed each time. What I see is five different images being generated during the stable diffusion pass, but then during the GFPGAN pass only the second image is processed -- five times! So I get five of the same image. The GFPGAN logging messages do refer to the correct seeds (the fixed starting seed followed by 4 randomly generated seeds), so the logic is off somewhere.

blessedcoolant commented:
> @blessedcoolant I've run into another problem. For benchmarking I'm running 5 iterations with the same starting seed each time. What I see is five different images being generated during the stable diffusion pass, but then during the GFPGAN pass only the second image is processed -- five times! So I get five of the same image. The GFPGAN logging messages do refer to the correct seeds (the fixed starting seed followed by 4 randomly generated seeds), so the logic is off somewhere.

Found the issue. Fixing it.

blessedcoolant commented:
@lstein Please check now. Fixed it.

blessedcoolant commented:
@lstein Did some minor cleanup after. The code was a bit dirty to my liking.

lstein commented:
Hi folks. I've done a series of benchmarks on my 32G GPU, and summary is here: https://docs.google.com/spreadsheets/d/1YXRlzp9TT9gRQMvbzKXW0xRjUI4bllHgfVcil0kS4oI/edit?usp=sharing

Take home messages:
1. @Oceanswave PRs run about 10% faster than @Blessedcoolant 's
2. Blessedcoolant's VRAM memory footprint is consistently lower than oceanswave. This was harder to measure, but ranged from 10%-20% depending on the task
3. When I pushed the system up to 832x1024 images, Blessedcoolant's implementation handled it (using 23.9G of VRAM), and Oceanswave consistently got CUDA OOM
4. I could not reproduce Issue #112 on either branch. This is a report of OOM errors occuring after several -n iterations, which suggested a memory leak of some sort.
5. When neither upscaling nor face fixing are requested, both branches perform equivalently. However, including --gfpgan in Oceanswave's branch does increase initialization time slightly and adds about 0.5G of VRAM utilization.

I think that users who want high quality faces and upscaling would rather trade off 10% runtime increase for 10% VRAM utilisation decrease. So my inclination is to bring in @blessedcoolant 's implementation and acknowledge both contributors for contributing the code to this working. I am so impressed with the improvements one can achieve with the combo of GFPGAN and Real-ESGAN!

I'll just wait for the afternoon (~4h) for any comments on the benchmarking and conclusions before hitting the merge button.

blessedcoolant commented:
> Hi folks. I've done a series of benchmarks on my 32G GPU, and summary is here: https://docs.google.com/spreadsheets/d/1YXRlzp9TT9gRQMvbzKXW0xRjUI4bllHgfVcil0kS4oI/edit?usp=sharing
> 
> Take home messages:
> 
> 1. @Oceanswave PRs run about 10% faster than @blessedcoolant 's
> 2. Blessedcoolant's VRAM memory footprint is consistently lower than oceanswave. This was harder to measure, but ranged from 10%-20% depending on the task
> 3. When I pushed the system up to 832x1024 images, Blessedcoolant's implementation handled it (using 23.9G of VRAM), and Oceanswave consistently got CUDA OOM
> 4. I could not reproduce Issue [GFPGAN executing fine for 1 image at a time, but OOM when using -n parameter, is there method to clear memory in between ? #112](https://github.com/lstein/stable-diffusion/issues/112) on either branch. This is a report of OOM errors occuring after several -n iterations, which suggested a memory leak of some sort.
> 5. When neither upscaling nor face fixing are requested, both branches perform equivalently. However, including --gfpgan in Oceanswave's branch does increase initialization time slightly and adds about 0.5G of VRAM utilization.
> 
> I think that users who want high quality faces and upscaling would rather trade off 10% runtime increase for 10% VRAM utilisation decrease. So my inclination is to bring in @blessedcoolant 's implementation and acknowledge both contributors for contributing the code to this working. I am so impressed with the improvements one can achieve with the combo of GFPGAN and Real-ESGAN!
> 
> I'll just wait for the afternoon (~4h) for any comments on the benchmarking and conclusions before hitting the merge button.

Thank you so much for taking the time to do this testing. Looking forward to the merge.

tildebyte commented:
@blessedcoolant; I apologize profusely for coming into this so late, but this PR **really** needs to have a `git rebase -i` run on it. The commit history for the repo as a whole is going to be messy and confusing if this is merged as is.

OTOH, the final merge decision is, as always, up to @lstein.

blessedcoolant commented:
> @blessedcoolant; I apologize profusely for coming into this so late, but this PR **really** needs to have a `git rebase -i` run on it. The commit history for the repo as a whole is going to be messy and confusing if this is merged as is.
> 
> OTOH, the final merge decision is, as always, up to @lstein.

I agree. It had a lot of changes made. I should have kept it cleaner but I was focused on the optimization. Sorry.

If @lstein wishes to, he can rebase this and pull it. If not, I can do a fresh PR with a cleaner commit history. Either works fine with you.

You could rebase it and make a PR that I can pull too if you have the time. I just popped out and won't be back for a few hours and don't want to hold up @lstein if he wishes to merge it before that.

tildebyte commented:
> You could rebase it

I really did consider this, but I'd be starting from 0 in understanding what's going on in here, which is, eh, "sub-optimal" at best when doing aggressive rebasing 😆

blessedcoolant commented:
> > You could rebase it
> 
> I really did consider this, but I'd be starting from 0 in understanding what's going on in here, which is, eh, "sub-optimal" at best when doing aggressive rebasing 😆

Here's all the list of changed files.

ldm\dream\pngwriter.py
ldm\gfpgan\gfpgan_tools.py
ldm\simplet2i.py
scripts\dream.py
scripts\preload_models.py
README.md

lstein commented:
Yes, I'm definitely going to rebase.

lstein commented:
@blessedcoolant I'm actually having a difficult time getting this to rebase cleanly. I wonder if you could rebase your commits and generate a clean PR against main? Thanks.

blessedcoolant commented:
> @blessedcoolant I'm actually having a difficult time getting this to rebase cleanly. I wonder if you could rebase your commits and generate a clean PR against main? Thanks.

Yep. I'm closing this PR. I'll raise a new clean one.

lstein commented:
Thanks so much.

blessedcoolant commented:
Picked up in #162
